src/main/java/com/google/javascript/gents/TypeConversionPass.java

```diff
@@ -87,50 +87,78 @@ public void visit(NodeTraversal t, Node n, Node parent) {
+          // still type aliases.
           bestJSDocInfo = NodeUtil.getBestJSDocInfo(n);
           if (bestJSDocInfo != null && bestJSDocInfo.hasTypedefType()) {
-            String interfaceName = n.getSecondChild().getString();
-            Node interfaceMember = Node.newString(Token.INTERFACE_MEMBERS, interfaceName);
-            typesToRename.put(n.getQualifiedName(), interfaceName);
-            typesToFilename.put(n.getQualifiedName(), n.getSourceFileName());
-            types.put(interfaceName, interfaceMember);
-            interfaceMember.setJSDocInfo(bestJSDocInfo);
-            Node interfaceNode = new Node(Token.INTERFACE, IR.empty(), IR.empty(), interfaceMember);
-            Node nameNode = Node.newString(Token.NAME, interfaceName);
-            nameNode.addChildToBack(interfaceNode);
-            Node exportNode = new Node(Token.EXPORT, new Node(Token.CONST, nameNode));
-            replaceExpressionOrAssignment(n, parent, exportNode);
+            Node typedefNode = bestJSDocInfo.getTypedefType().getRoot();
+            if (containsSimpleTypes(typedefNode)) {
+              createTypeAlias(n, parent);
+            } else {
+              String interfaceName = n.getSecondChild().getString();
+              Node interfaceMember = Node.newString(Token.INTERFACE_MEMBERS, interfaceName);
+              typesToRename.put(n.getQualifiedName(), interfaceName);
+              typesToFilename.put(n.getQualifiedName(), n.getSourceFileName());
+              types.put(interfaceName, interfaceMember);
+              interfaceMember.setJSDocInfo(bestJSDocInfo);
+              Node interfaceNode =
+                  new Node(Token.INTERFACE, IR.empty(), IR.empty(), interfaceMember);
+              Node nameNode = Node.newString(Token.NAME, interfaceName);
+              nameNode.addChildToBack(interfaceNode);
+              Node exportNode = new Node(Token.EXPORT, new Node(Token.CONST, nameNode));
+              replaceExpressionOrAssignment(n, parent, exportNode);
+            }
           }
           break;
         case NAME:
         case VAR:
         case LET:
         case CONST:
-          // Creates a type alias
-          bestJSDocInfo = NodeUtil.getBestJSDocInfo(n);
-          if (bestJSDocInfo != null && bestJSDocInfo.hasTypedefType()) {
-            String name;
-            if (n.getToken() == Token.NAME) {
-              name = n.getString();
-            } else {
-              name = n.getFirstChild().getString();
-            }
-            Node typeDef = Node.newString(Token.TYPE_ALIAS, name);
-            types.put(name, typeDef);
-            typeDef.setJSDocInfo(bestJSDocInfo);
-            replaceExpressionOrAssignment(n, parent, typeDef);
-          }
+          createTypeAlias(n, parent);
           break;
         default:
           break;
       }
     }
 
+    private boolean containsSimpleTypes(Node typedefNode) {
```

Comment:
This should be the other way around – we should whitelist specifically the pattern we support, and reject everything else.

Our code depending on this only supports a specific pattern, so we should detect the specific pattern and handle it. I.e. this should be called `containsRecordType` and you should invert the condition. I think the logic is fine though.